src/atom.rs

```diff
@@ -0,0 +1,166 @@
+}
+
+pub struct AtomIter<'a> {
+    bytes: &'a [u8],
+    offset: usize,
+}
+
+// https://tools.ietf.org/html/rfc3629
+#[cfg_attr(rustfmt, rustfmt_skip)]
+static UTF8_CHAR_WIDTH: [u8; 256] = [
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // 0x1F
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // 0x3F
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // 0x5F
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // 0x7F
+    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, // 0x9F
+    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, // 0xBF
+    0,0,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
+    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2, // 0xDF
+    3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3, // 0xEF
+    4,4,4,4,4,0,0,0,0,0,0,0,0,0,0,0, // 0xFF
+];
+
+impl<'a> Iterator for AtomIter<'a> {
+    type Item = Atom<'a>;
+
+    fn next(&mut self) -> Option<Atom<'a>> {
+        // No more characters.
+        if self.offset >= self.bytes.len() {
+            return None;
+        }
+
+        let x = self.bytes[self.offset];
+
+        // Underscore or Number.
+        match x {
+            // Underscore.
+            b'_' => {
+                self.offset += 1;
+
+                let mut n = 1;
+                while self.offset < self.bytes.len() {
+                    match self.bytes[self.offset] {
+                        b'_' => {
+                            self.offset += 1;
+                            n += 1;
+                        }
+                        _ => break,
+                    }
+                }
+
+                Some(Atom::Underscore(n))
+            }
+            // Number.
+            b'0'..=b'9' => {
+                let start = self.offset;
+
+                self.offset += 1;
+                while self.offset < self.bytes.len() {
+                    match self.bytes[self.offset] {
+                        b'0'..=b'9' => self.offset += 1,
+                        _ => break,
+                    }
+                }
+
+                // This is safe since we've verified it's valid utf8.
+                let s = &self.bytes[start..self.offset];
+                let s = unsafe { from_utf8_unchecked(s) };
+
+                let int = match s.parse::<u64>() {
+                    Ok(n) => n,
+                    _ => unreachable!("expected integer"),
+                };
+
+                Some(Atom::Number(int))
+            }
+            // Don't care.
+            _ => {
+                let start = self.offset;
+
+                self.offset += 1;
+                while self.offset < self.bytes.len() {
+                    match self.bytes[self.offset] {
+                        b'_' | b'0'..=b'9' => break,
+                        b => {
+                            // Use a copy of the UTF8_CHAR_WIDTH array from core::str to avoid
+                            // relying on unstable internals.
```

Comment:
I don't want to do this. Could you find a way to implement this without copying a utf8 table and without unsafe code?

Reply:
We can always go on the assumption that the bytes are valid and not bother checking them.